Update overview + reorganize Rust tutorial (#739)
CryptozombiesHQ/cryptozombie-lessons: en/21/07.md

---
title: Function Modifiers
actions: ['checkAnswer', 'hints']
material:
  editor:
    language: rust

>     startingCode: |

Reviewer: ```suggestion
    startingCode:
```

      "lib.rs": |
        #![no_std]

        multiversx_sc::imports!();
        multiversx_sc::derive_imports!();

        #[derive(NestedEncode, NestedDecode, TopEncode, TopDecode, TypeAbi)]
        pub struct Zombie<M: ManagedTypeApi> {
            name: ManagedBuffer<M>,
            dna: u64,
        }

        #[multiversx_sc::contract]
        pub trait ZombiesContract {

          #[init]
          fn init(&self) {
            self.dna_digits().set(16u8);
            self.zombies_count().set(1usize);
          }

          fn create_zombie(&self, name: ManagedBuffer, dna: u64) {
              self.zombies_count().update(|id| {
                self.zombies(id).set(Zombie { name, dna });
                *id +=1;
              });
          }

          #[storage_mapper("dna_digits")]
          fn dna_digits(&self) -> SingleValueMapper<u8>;

          #[storage_mapper("zombies_count")]
          fn zombies_count(&self) -> SingleValueMapper<usize>;
          
          #[storage_mapper("zombies")]
          fn zombies(&self, id: &usize) -> SingleValueMapper<Zombie<Self::Api>>;
        }

>     answer: >

Reviewer: ```suggestion
    answer: |
```

      multiversx_sc::derive_imports!();

      #[derive(NestedEncode, NestedDecode, TopEncode, TopDecode, TypeAbi)]
      pub struct Zombie<M: ManagedTypeApi> {
          name: ManagedBuffer<M>,
          dna: u64,
      }

      #[multiversx_sc::contract]
      pub trait ZombiesContract {

        #[init]
        fn init(&self) {
          self.dna_digits().set(16u8);
          self.zombies_count().set(1usize);
        }

        fn create_zombie(&self, name: ManagedBuffer, dna: u64) {
            self.zombies_count().update(|id| {
              self.zombies(id).set(Zombie { name, dna });
              *id +=1;
            });
        }

        #[view]
        fn generate_random_dna(&self) -> u64 {
        }

        #[storage_mapper("dna_digits")]
        fn dna_digits(&self) -> SingleValueMapper<u8>;

        #[storage_mapper("zombies_count")]
        fn zombies_count(&self) -> SingleValueMapper<usize>;

        #[view]
        #[storage_mapper("zombies")]
        fn zombies(&self, id: &usize) -> SingleValueMapper<Zombie<Self::Api>>;
      }
---

In this chapter, we're going to learn about function **_return values_**, and function modifiers.

## Return Values

> To return a value from a function, the declaration looks like this:

Reviewer: ```suggestion
The following example creates a function named `say_hello` that returns "What's up dog":
```

> ```

Reviewer: ```suggestion
```rust
```

fn say_hello() -> ManagedBuffer {
  let greeting = ManagedBuffer::from(b"What's up dog");
  return greeting;
}

```
Alternativelly you can leave the returned element as the last line of the function without the delimiter `;` having the same effect as the classic format:

> ```

Reviewer: ```suggestion
```rust
```

function say_hello() -> ManagedBuffer {
  ManagedBuffer::from(b"What's up dog")
}

```

In Rust, the function declaration contains the type of the return value (in this case `ManagedBuffer`) indicated by `->`

## Function modifiers

> The above function doesn't actually change state in Rust — e.g. it doesn't change any values or write anything.

Reviewer: ```suggestion
The example function above doesn't change the state in Rust, meaning it doesn't change any values or write anything.
```
